### PR TITLE
Add ##__VA_ARGS__ and __VA_OPT__

### PR DESCRIPTION
**Pull request by BadRyuner**
Examples for the tests I took from here
https://en.cppreference.com/w/c/preprocessor/replace#Function-like_macros
The code passed all the tests

### CODE REVIEW

File: Cesium.Preprocessor/MacroExpansionEngine.cs [hunk 1]
```diff
@@ -268,6 +272,18 @@ IEnumerable<IToken<CPreprocessorTokenType>> ClearSpaceBuffer()
         }
         yield break;
 
+        void PoofComma()
+        {
+            foreach (var tok in lexer.EnumerateInReverse()) // look at previous tokens
+                if (tok.Text == ",") // is that a comma?
+                {
+                    MemoryMarshal.GetReference(tok.Text.AsSpan()) = ' '; // poof and there`s no more comma
```
Comment (ForNeVeR) [inline Cesium.Preprocessor/MacroExpansionEngine.cs:280]: This modifies a `string` in-place which may have consequences. Please avoid doing that. All the compiler structures, including AST, lexer data and whatnot are immutable, not even starting about `string`.

File: Cesium.Preprocessor/MacroExpansionEngine.cs [hunk 2]
```diff
@@ -10,6 +11,8 @@ namespace Cesium.Preprocessor;
 
 public class MacroExpansionEngine(IWarningProcessor warningProcessor, IMacroContext macroContext)
 {
+    private string __VA_OPT__Rule = ",";
```
Comment (ForNeVeR) [inline Cesium.Preprocessor/MacroExpansionEngine.cs:14]: I am not quite sure about this state introduced into the engine. Note that it's reused between different preprocessor calls; it wasn't meant to have any state. And here you introduce a variable that may change while expanding `__VA_OPT__`. What if we need to expand it twice? What if we need to expand it in different macros?

I'd suggest avoid relying on this kind of state.

**Review by ForNeVeR — CHANGES_REQUESTED** [review]

File: Cesium.Preprocessor/MacroExpansionEngine.cs (hunk 1 above)
Comment (BadRyuner) [reply Cesium.Preprocessor/MacroExpansionEngine.cs:280]: I don't see any other way to remove the previous token if we create a new collection via yield return that we don't have access to.

Comment (ForNeVeR) [reply Cesium.Preprocessor/MacroExpansionEngine.cs:280]: Alright, I'll add this to my queue to rearchitecture the corresponding part.

File: Cesium.Preprocessor/MacroExpansionEngine.cs (hunk 2 above)
Comment (BadRyuner) [reply Cesium.Preprocessor/MacroExpansionEngine.cs:14]: Then add VA_OPT as a field in MacroArguments?

**Pull request closed without merging**